canlink/busmanager.go

```diff
@@ -0,0 +1,242 @@
+	isRunning bool
+	mu        sync.Mutex
+}
+
+// NewBusManager returns a BusManager object.
+//
+// The network connection is injected into the BusManager
+// and provides the interface for a single bus.
+//
+// See usage example.
+func NewBusManager(l *zap.Logger, conn *net.Conn) *BusManager {
+	busManager := &BusManager{
+		l: l.Named("Bus Manager"),
+
+		broadcastChan: make(map[Handler]chan TimestampedFrame),
+		incomingChan:  make(chan TimestampedFrame, _channelBufferLength),
+
+		receiver:    socketcan.NewReceiver(*conn),
+		transmitter: socketcan.NewTransmitter(*conn),
+	}
+
+	return busManager
+}
+
+// Register a Handler with the BusManager.
+//
+// Register returns two channels for the specified Handler.
+// The broadcast channel is a stream of traffic received from
+// the bus. The incoming channel enables the Handler to write
+// a frame to the bus.
+//
+// The channels operate on a TimestampedFrame object.
+func (b *BusManager) Register(
+	handler Handler,
+) (chan TimestampedFrame, chan TimestampedFrame) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	b.l.Info("registering handler")
+
+	subscription := make(chan TimestampedFrame, _channelBufferLength)
+	b.broadcastChan[handler] = subscription
+
+	b.l.Info("registered handler")
+
+	return subscription, b.incomingChan
+}
+
+// Unregister a Handler from the BusManager.
+//
+// Deletes the broadcast channel that was previously
+// provided from BusManager.
+func (b *BusManager) Unregister(handler *Handler) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	delete(b.broadcastChan, *handler)
+}
+
+// Start the traffic broadcast and incoming frame listener
+// for each of the registered handlers.
+//
+// The broadcast stream will begin. If handlers cannot keep up
+// with the broadcast, frames for that handler will be dropped.
+//
+// The handlers can now send CAN frames to the BusManager to be
+// transmitted out onto the bus.
+func (b *BusManager) Start(ctx context.Context) {
+	b.l.Info("start broadcast and incoming")
+
+	if b.isRunning {
+		return
+	}
+
+	b.stop = make(chan struct{})
+
+	go b.broadcast(ctx)
+	go b.processIncoming(ctx)
+
+	b.isRunning = true
+}
+
+// Stop the traffic broadcast and incoming frame listener.
+//
+// Preserves registered handlers and their assosciated channels.
+func (b *BusManager) Stop() {
+	b.l.Info("stop broadcast and incoming")
+
+	if !b.isRunning {
+		return
+	}
+
+	close(b.stop)
+	b.isRunning = false
+}
+
+// Close cleans up the bus network connection.
+func (b *BusManager) Close(conn net.Conn) error {
+	b.l.Info("closing socketcan receiver and transmitter")
+
+	if b.isRunning {
+		return errors.New("cannot close active bus manager")
+	}
+
+	b.receiver.Close()
+	b.transmitter.Close()
+
+	return nil
+}
+
+func (b *BusManager) broadcast(ctx context.Context) {
+	timeFrame := TimestampedFrame{}
+
+	for b.receiver.Receive() {
+		timeFrame.Frame = b.receiver.Frame()
+		timeFrame.Time = time.Now()
+
+		b.mu.Lock()
+
+		for handler, ch := range b.broadcastChan {
```

Comment:
where does the message get passed to the handler?

Reply:
It never gets passed to the handler, rather it gets sent into its associated broadcast channel (see line 213), the handler in this loop is only used in logging it seems.

Reply:
@langei How's this work and where do I define my message handling behaviour? I thought the purpose of the BusManager was to send the message to each Handler.